whisper.cpp

```diff
@@ -1708,7 +1705,10 @@ static struct ggml_cgraph * whisper_build_graph_encoder(
 
     ggml_cgraph * gf = ggml_new_graph_custom(ctx0, WHISPER_MAX_NODES, false);
 
-    struct ggml_tensor * cur = ggml_view_tensor(ctx0, wstate.embd_conv);
+    // TODO: this still triggers the assert:
+    //struct ggml_tensor * cur = ggml_view_tensor(ctx0, wstate.embd_conv);
```

Comment:
With the `wstate.embd_conv` now pre-allocated, this view still triggers the assert:

```
Assertion failed: (tensor_alloc->offset == SIZE_MAX), function ggml_gallocr_init_tensor, file ggml-alloc.c, line 739.
```

Not sure why

Reply:
The code that stores the allocation location of leafs didn't take into account pre-allocated leafs. This should fix it:
```diff
diff --git a/ggml-alloc.c b/ggml-alloc.c
index 8ac1d3e..60b86c2 100644
--- a/ggml-alloc.c
+++ b/ggml-alloc.c
@@ -701,8 +701,13 @@ bool ggml_gallocr_reserve_n(ggml_gallocr_t galloc, struct ggml_cgraph * graph, c
         struct ggml_tensor * leaf = graph->leafs[i];
         struct hash_node * hn = ggml_gallocr_hash_get(galloc, leaf);
         galloc->leaf_allocs[i].buffer_id = hn->buffer_id;
-        galloc->leaf_allocs[i].leaf.offset = hn->offset;
-        galloc->leaf_allocs[i].leaf.size_max = ggml_backend_buft_get_alloc_size(galloc->bufts[hn->buffer_id], leaf);
+        if (leaf->view_src || leaf->data) {
+            galloc->leaf_allocs[i].leaf.offset = SIZE_MAX;
+            galloc->leaf_allocs[i].leaf.size_max = 0;
+        } else {
+            galloc->leaf_allocs[i].leaf.offset = hn->offset;
+            galloc->leaf_allocs[i].leaf.size_max = ggml_backend_buft_get_alloc_size(galloc->bufts[hn->buffer_id], leaf);
+        }
     }

     // reallocate buffers if needed
```

Reply:
With just this change applied on `master` the code also works (i.e. without pre-allocating the tensors). But if I understand correctly, it is technically not correct because nothing guarantees that the data in these tensors would not be overwritten by some ops. Since they are currently at the end of the computation graphs it seems to produce correct results, but the concern is that this is not very future-proof if we expand the graphs in the future - is that correct?

What if I use `ggml_set_output(ctx0, wstate.embd_conv);` in `whisper_build_graph_conv()` to guarantee that the data would not be overwritten?

Reply:
Yes, with `ggml_set_output` it would guarantee that the tensor is never overwritten.

Reply:
Would pre-allocating the tensors as proposed in this PR have any advantage over the `ggml_set_output` option? Since with this PR we now have to perform extra copy of the data between the graph calls, which otherwise is not needed.

Reply:
I don't think there would be any significant advantage. Maybe it would using `ggml_gallocr_reserve` here instead of `ggml_gallocr_alloc_graph`, but the difference would be minimal:

https://github.com/ggerganov/whisper.cpp/blob/66df44b0b7f2b1671e59379ad019df3b71b2547c/whisper.cpp#L496-L502